packages/flutter/lib/src/services/text_editing_value.dart

```diff
@@ -0,0 +1,444 @@
+      case 0x2002: // en space
+      case 0x2003: // em space
+      case 0x2004: // three-per-em space
+      case 0x2005: // four-er-em space
+      case 0x2006: // six-per-em space
+      case 0x2007: // figure space
+      case 0x2008: // punctuation space
+      case 0x2009: // thin space
+      case 0x200A: // hair space
+      case 0x202F: // narrow no-break space
+      case 0x205F: // medium mathematical space
+      case 0x3000: // ideographic space
+        break;
+      default:
+        return false;
+    }
+    return true;
+  }
+
+  /// Returns the index into the string of the next character boundary after the
+  /// given index.
+  ///
+  /// The character boundary is determined by the characters package, so
+  /// surrogate pairs and extended grapheme clusters are considered.
+  ///
+  /// The index must be between 0 and string.length, inclusive. If given
+  /// string.length, string.length is returned.
+  ///
+  /// Setting includeWhitespace to false will only return the index of non-space
+  /// characters.
+  static int nextCharacter(int index, String string, [bool includeWhitespace = true]) {
+    assert(index >= 0 && index <= string.length);
+    if (index == string.length) {
+      return string.length;
+    }
+
+    int count = 0;
+    final Characters remaining = string.characters.skipWhile((String currentString) {
+      if (count <= index) {
+        count += currentString.length;
+        return true;
+      }
+      if (includeWhitespace) {
+        return false;
+      }
+      return isWhitespace(currentString.codeUnitAt(0));
+    });
+    return string.length - remaining.toString().length;
+  }
+
+  /// Returns the index into the string of the previous character boundary
+  /// before the given index.
+  ///
+  /// The character boundary is determined by the characters package, so
+  /// surrogate pairs and extended grapheme clusters are considered.
+  ///
+  /// The index must be between 0 and string.length, inclusive. If index is 0,
+  /// 0 will be returned.
+  ///
+  /// Setting includeWhitespace to false will only return the index of non-space
+  /// characters.
+  static int previousCharacter(int index, String string, [bool includeWhitespace = true]) {
+    assert(index >= 0 && index <= string.length);
+    if (index == 0) {
+      return 0;
+    }
+
+    int count = 0;
+    int? lastNonWhitespace;
+    for (final String currentString in string.characters) {
+      if (!includeWhitespace &&
+          !isWhitespace(currentString.characters.first.codeUnitAt(0))) {
+        lastNonWhitespace = count;
+      }
+      if (count + currentString.length >= index) {
+        return includeWhitespace ? count : lastNonWhitespace ?? 0;
+      }
+      count += currentString.length;
+    }
+    return 0;
+  }
+
+  /// Return the offset at the end of the nearest word to the right of the given
+  /// offset.
+  static int getRightByWord(String text, TextMetrics textMetrics, int offset, [bool includeWhitespace = true]) {
```

Comment:
I actually had already moved this to TextEditingActionTarget as a private instance method that uses TextMetrics.getWordPosition, but I had forgotten to delete this static method here.  Thanks for catching this.  Let me know if you like it on TextEditingActionTarget or if you think it would be better on TextMetrics.

Regarding RTL, you're right, I think we have a lot of missing support for this.  I've created an issue (https://github.com/flutter/flutter/issues/88126) and I'll add TODOs for it here.